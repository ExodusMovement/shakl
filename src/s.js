import React from 'react';

import flatten from './flatten';

const s = (Comp, { name, multi, ...opts } = {}) => style => {
  const S = React.forwardRef(({ children, ...props }, ref) => {
    const { attrs = {}, comp, child } = opts;

    const styles = multi ? style : { style };

    return React.createElement(
      comp || Comp,
      {
        ref,
        ...attrs,
        ...props,
        ...Object.keys(styles).reduce(
          (obj, prop) => ({
            ...obj,
            [prop]: flatten([
              attrs[prop],
              typeof styles[prop] === 'function'
                ? styles[prop](props)
                : styles[prop],
              props[prop]
            ])
          }),
          {}
        )
      },
      child ? React.createElement(child, {}, children) : children
    );
  });

  S.displayName = name || `styled(${Comp.displayName || Comp.name})`;

  S.extend = more => s(S, { name })(more);
  S.withProps = attrs => s(S, { name, attrs })();
  S.withComponent = comp => s(S, { name, comp })(style);
  S.withChild = child => s(S, { name, child })();

  return S;
};

export default s;
