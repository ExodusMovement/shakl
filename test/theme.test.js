import React from 'react';

import { create as r } from 'react-test-renderer';

import s, { ThemeProvider } from '../src';

it('creates a styled component with theme', () => {
  const style = { padding: 20 };

  const Foo = s
    .Text(({ theme }) => ({
      backgroundColor: theme.backgroundColor,
      color: theme.color,
      ...style
    }))
    .withTheme();

  const lightTheme = { backgroundColor: 'white', color: 'black' };
  const darkTheme = { backgroundColor: 'black', color: 'white' };

  const ThemedFoo = props => (
    <ThemeProvider {...props}>
      <Foo />
    </ThemeProvider>
  );

  const light = r(<ThemedFoo theme={lightTheme} />).toJSON();
  const dark = r(<ThemedFoo theme={darkTheme} />).toJSON();

  expect(light.props.style).toEqual({ ...lightTheme, ...style });
  expect(dark.props.style).toEqual({ ...darkTheme, ...style });

  expect(light).toMatchSnapshot();
  expect(dark).toMatchSnapshot();
});
