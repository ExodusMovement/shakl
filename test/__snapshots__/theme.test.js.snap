// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`creates a styled component with theme 1`] = `
<Text
  accessible={true}
  allowFontScaling={true}
  ellipsizeMode="tail"
  style={
    Object {
      "backgroundColor": "white",
      "color": "black",
      "padding": 20,
    }
  }
  theme={
    Object {
      "backgroundColor": "white",
      "color": "black",
    }
  }
/>
`;

exports[`creates a styled component with theme 2`] = `
<Text
  accessible={true}
  allowFontScaling={true}
  ellipsizeMode="tail"
  style={
    Object {
      "backgroundColor": "black",
      "color": "white",
      "padding": 20,
    }
  }
  theme={
    Object {
      "backgroundColor": "black",
      "color": "white",
    }
  }
/>
`;
